File: array/0121-best-time-to-buy-and-sell-stock/solution.cpp
// 121. Best Time to Buy and Sell Stock
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Easy | C++ | Accepted 2026-08-04
// Runtime 6 ms | Memory 97.3 MB

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxx = -1;
        int minn = 1000000;
        int maxP = 0;
        for(int i = 0; i<prices.size(); i++)
        {
            if(minn!=1000000 && maxx!=-1)
            {
                maxP = max(maxP, maxx-minn);
            }
            maxx = max(maxx, prices[i]);
            if(prices[i]<minn)
            {
            minn = min(minn, prices[i]);
            maxx = -1;
            }
        }
        return max(maxx-minn, maxP);
    }
};

File: array/0121-best-time-to-buy-and-sell-stock/solution.java
// 121. Best Time to Buy and Sell Stock
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Easy | Java | Accepted 2025-12-19
// Runtime 2 ms | Memory 94.7 MB

class Solution {
    public int maxProfit(int[] prices) {
        int start = prices[0];
        int maxProf = 0;
        int ans = 0;
        for(int i = 1; i<prices.length; i++)
        {
            if(prices[i]-start>maxProf)
            {
                maxProf = prices[i]-start;
            }
            if(prices[i]-start<0)
            {
                start = prices[i]; 
            }
            ans = Math.max(maxProf, 0);
        }
        return ans;
    }
}
